Untrusted data flow to external API

owen-mc opened the pull request:
I have based this on the java version (cf. https://github.com/github/codeql/pull/3938).

owen-mc commented:
@lcartey if you have time you could try out this go version of "Untrusted data flow to external API" and give feedback before it's merged.

smowton commented on ql/src/Security/CWE-020/ExternalAPISinkExample.go:
```diff
@@ -0,0 +1,20 @@
+			// BAD: a request parameter is incorporated without validation into the response
+			fmt.Fprintf(w, "%q is an unknown user", username)
+		} else {
+			// TODO: do something exciting
```
?

owen-mc replied:
This test is copied from ReflectedXSS.

smowton replied:
Weird. Suggest replacing this with `// Handle successful login` in both cases; at the moment it's ambiguous whether we forgot to finish writing to test

smowton commented on ql/src/Security/CWE-020/ExternalAPIsUsedWithUntrustedData.qhelp:
```diff
@@ -0,0 +1,49 @@
+may be relevant for security analysis of this application.</p>
+
+<p>An external API is defined as a call to a function that is not defined in the source code and is not
+modeled as a taint step in the default taint library. Calls made in test files are also excluded.
```
```suggestion
modeled as a taint step in the default taint library. Calls made in test files are excluded.
```

smowton replied:
NB. most of these look like they apply to other qhelp files. Did you already ping the docs team about whether we can reduce duplication by referring to other qhelp files?

owen-mc replied:
They said it wasn't possible

owen-mc replied:
I have applied each comment on a qhelp file to all the qhelp files that it makes sense for

smowton commented on ql/src/Security/CWE-020/ExternalAPIsUsedWithUntrustedData.qhelp:
```diff
@@ -0,0 +1,49 @@
+<p>An external API is defined as a call to a function that is not defined in the source code and is not
+modeled as a taint step in the default taint library. Calls made in test files are also excluded.
+External APIs may be from the  Go standard library, third party dependencies or from internal dependencies.
+The query will report the fully qualified method name, along with either <code>[param x]</code>,
```
```suggestion
The query will report the fully-qualified method name, along with either <code>[param x]</code>,
```

smowton commented on ql/src/Security/CWE-020/ExternalAPIsUsedWithUntrustedData.qhelp:
```diff
@@ -0,0 +1,49 @@
+<ul>
+  <li>If the result highlights a known sink, no action is required.</li>
+  <li>If the result highlights an unknown sink for a problem, then add modeling for the sink to the relevant query.</li>
+  <li>If the result represents a call to an external API which transfers taint, add the appropriate modeling, and 
```
```suggestion
  <li>If the result represents a call to an external API which transfers taint (for example, from a parameter to its return value), add the appropriate modeling, and 
```

smowton commented on ql/src/Security/CWE-020/UntrustedDataToExternalAPI.qhelp:
```diff
@@ -0,0 +1,60 @@
+<overview>
+<p>Using unsanitized untrusted data in an external API can cause a variety of security issues. This query reports 
+external APIs that use untrusted data. The results have very little filtering so that you can audit almost all
+examples. The query provides data for security reviews of the application and you can also use it to identify
```
```suggestion
examples. The query provides data for security reviews of the application. It can also be used to identify
```

smowton commented on ql/src/Security/CWE-020/ExternalAPIsUsedWithUntrustedData.qhelp:
```diff
@@ -0,0 +1,49 @@
+<p>If the query were to return the API <code>fmt.Sprintf [param 1]</code>, then this should be
+reviewed as a possible taint step, because tainted data would flow from the 1st argument to the return value
+of the call.</p>
+
```
Missing `sample` tags in this one?

owen-mc replied:
I see. I copied the java one, which doesn't use `sample` tags, but the functions that are in the go examples aren't http-specific enough for the comment about XSS to make sense. I'll put them in.

smowton commented on ql/src/semmle/go/security/ExternalAPIs.qll:
```diff
@@ -0,0 +1,216 @@
+      this = call.getReceiver() and
+      i = -1 and
+      (
+        call.getTarget().getNumResult() >= 2
```
nontrivial meaning anything other than a single boolean? What is this for?

owen-mc replied:
I think "nontrivial" also means non-empty, i.e. there is a return value. It might be clearer to rephrase this and the following lines as
```not call.getTarget().getNumResult() = 0 and not (call.getTarget().getNumResult() = 1 and call.getTarget().getResultType(0) instanceof BoolType)```

smowton replied:
Right, but why do we care? `void`-returning methods seem like as valid an external call as any other?

owen-mc replied:
Apparently this was added for the java version because there were lots of those functions and they weren't very useful. I have added a commit removing this restriction. I can always reinstate it if we find the same as them.

smowton commented on ql/src/semmle/go/security/ExternalAPIs.qll:
```diff
@@ -0,0 +1,216 @@
+    ) and
+    // Not defined in the code that is being analysed
+    not exists(call.getACallee().getBody()) and
+    // Not a function pointer, unless it's declared in a package
```
```suggestion
    // Not a function pointer, unless it's declared at package scope
```
(everything is in a package)

smowton commented on ql/src/semmle/go/security/ExternalAPIs.qll:
```diff
@@ -0,0 +1,216 @@
+TaintTracking::FunctionModel getAMethodModelInPackage(Package p) {
+  p = result.getPackage() and
+  result instanceof Method and
+  result.getName() != "String" and
```
Comment why the String method is particularly excluded? I can only half remember myself

smowton commented on ql/src/semmle/go/security/ExternalAPIs.qll:
```diff
@@ -0,0 +1,216 @@
+  n instanceof CleartextLogging::Sink
+}
+
+/** A node representing data being passed to an external API. */
```
```suggestion
/** A node representing data being passed to an unknown external API. */
```

smowton reviewed (approved):
Some minor things to attend to, mostly lgtm

lcartey commented:
@rvermeulen I know you wrote something similar recently, do you want to take a look?

shati-patel commented on ql/src/Security/CWE-020/UntrustedDataToUnknownExternalAPI.qhelp:
```diff
@@ -0,0 +1,61 @@
+
+<p>An external API is defined as a call to a function that is not defined in the source code and is not
+modeled as a taint step in the default taint library. Calls made in test files are excluded.
+External APIs may be from the Go standard library, third-party dependencies or from internal dependencies.
```
```suggestion
External APIs may be from the Go standard library, third-party dependencies, or from internal dependencies.
```

shati-patel commented on ql/src/Security/CWE-020/UntrustedDataToUnknownExternalAPI.qhelp:
```diff
@@ -0,0 +1,61 @@
+The query reports uses of untrusted data in either the receiver or as one of the arguments of external APIs.</p>
+
+<p>An external API is considered unknown if it is not in a package which has already been modeled, it is not
+a sink for an existing query and it is not in a list of external APIs which have been examined and determined
```
```suggestion
a sink for an existing query, and it is not in a list of external APIs which have been examined and determined
```

shati-patel commented on ql/src/Security/CWE-020/UntrustedDataToUnknownExternalAPI.qhelp:
```diff
@@ -0,0 +1,61 @@
+
+<sample src="ExternalAPISinkExample.go" />
+
+<p>This is an XSS sink. The XSS query should therefore be reviewed to confirm that this sink is appropriately modeled,
```
```suggestion
<p>This is an XSS sink. The "Reflected cross-site scripting" query (<code>go/reflected-xss</code>) should therefore be reviewed to confirm that this sink is appropriately modeled,
```

(Same as above)

shati-patel commented on ql/src/Security/CWE-020/UntrustedDataToUnknownExternalAPI.qhelp:
```diff
@@ -0,0 +1,61 @@
+re-run the query to determine what additional results might be found. In this example, an SQL injection vulnerability
+would be reported.</p>
+
+<p>Note that both examples are correctly handled by the standard taint tracking library, the reflected XSS query and SQL
+injection query.</p>
```
```suggestion
<p>Note that both examples are correctly handled by the standard taint tracking library,
the "Reflected cross-site scripting" query (<code>go/reflected-xss</code>),
and the "Database query built from user-controlled sources" query (<code>go/sql-injection</code>).</p>
```

(Same as above)

shati-patel reviewed (commented):
Looks good! I've added relevant suggestions to all the places they're repeated, so apologies for the duplicated comments 🙈 

There aren't any change notes yet, could you add them before this is merged?

owen-mc commented on ql/src/Security/CWE-020/ExternalAPIsUsedWithUntrustedData.ql:
```diff
@@ -0,0 +1,17 @@
+import semmle.go.security.ExternalAPIs
+
+from ExternalAPIUsedWithUntrustedData externalAPI
+select externalAPI, count(externalAPI.getUntrustedDataNode()) as numberOfUses,
```
This is a table query, so the results are put into a table. Looking at other table queries, it seems to be the norm that they don't have a string. Note that the variable names `externalAPI`, `numberOfUses`, and `numberOfUntrustedSources` become the column headings, so there is some explanation of what's going on.

shati-patel replied:
That makes sense then, thanks! Didn't realise about the table and the headings 😃

owen-mc commented:
I have addressed all review comments. I will run a distcompare job over the weekend as there are one or two library changes which will affect other queries (mainly adding more testing frameworks).

owen-mc commented:
The distcompare job looks fine. No changes in what results we find. The timings are about the same - certainly within expected variation.
